src/test/lesson_4/FractionalWrapper.t.sol

```diff
@@ -0,0 +1,210 @@
+        address indexed owner, 
+        uint256 assets, 
+        uint256 shares
+    );
+
+    event Withdraw(
+        address indexed caller, 
+        address indexed receiver, 
+        address indexed owner, 
+        uint256 assets, 
+        uint256 shares
+    );
+
+    address user;
+
+    function setUp() public virtual {
+        token = new FailedTransfers();
+        wrapper = new FractionalWrapper(token, 20 * (10**26), "Wrapped Coil", "WCOIL");
+        user = address(1);
+        vm.prank(user);
+        token.mint(user, 10**18);
+    }
+}
+
+abstract contract WithTokens is ZeroState {
+    function setUp() public override {
+        super.setUp();
+        vm.startPrank(user);
+        token.approve(address(wrapper), 10**18);
+        wrapper.deposit(10**18, user);
+        vm.stopPrank();
+    }
+}
+
+contract FractionalWrapperTest is ZeroState {
+    function testCorrectAsset() public {
+        console.log("Contains the correct asset as underlying");
+        assertEq(address(token), address(wrapper.asset()));
+    }
+
+    function testCorrectExchangeRate() public {
+        console.log("Has correct exchange rate");
+        assertEq(wrapper.exchangeRate(), 20 * (10**26));
+    }
+
+    function testTotalAssets() public {
+        console.log("Returns correct amount of total assets");
+        assertEq(wrapper.totalAssets(), 0);
+    }
+
+    function testConvertToShares() public {
+        console.log("Converts the amounts of assets successfully");
+        assertEq(wrapper.convertToShares(1), 2);
+        assertEq(wrapper.convertToShares(2), 4);
+        assertEq(wrapper.convertToShares(7), 14);
+        assertEq(wrapper.convertToShares(10), 20);
+        assertEq(wrapper.convertToShares(100), 200);
+    }
+
+    function testConvertToAssets() public {
+        console.log("Converts the amounts of shares successfully");
+        assertEq(wrapper.convertToAssets(1), 0);
+        assertEq(wrapper.convertToAssets(10), 5);
+        assertEq(wrapper.convertToAssets(70), 35);
+        assertEq(wrapper.convertToAssets(100), 50);
+        assertEq(wrapper.convertToAssets(200), 100);
+    }
+
+    function testMaxDeposit(address receiver) public {
+        console.log("Retrieves max deposit amount successfully");
+        assertEq(wrapper.maxDeposit(receiver), type(uint256).max);
+    }
+
+    function testPreviewDeposit(uint256 assets) public {
+        console.log("Gives correct amount of shares");
+        vm.assume(assets <= 1e27);
+        assertEq(wrapper.previewDeposit(assets), wrapper.convertToShares(assets));
+    }
+
+    function testDeposit(address receiver) public {
+        console.log("Deposits tokens successfully");
+        vm.startPrank(user);
+        token.approve(address(wrapper), 10);
+        vm.expectEmit(true, true, true, true);
+        emit Deposit(
+            user,
+            receiver, 
+            10, 
+            wrapper.convertToShares(10)
+        );
+        wrapper.deposit(10, receiver);
```

Comment:
The return value is not being tested.

Reply:
Do you mean like my `assertEq` statements in `testWithdraw` and `testRedeem`?

Reply:
I mean:
```
uint256 shares = wrapper.deposit(10, receiver);

assertEq(shares, ...)
```

I've messed up in the past by having a function return something wrong, and if you don't test it you might not find it until much later, because the frontend rarely will use the return values. It's only when someone (maybe yourself) decides to integrate another smart contract that you find out.

Reply:
I understand.